### ISSUE TITLE
Most `KnownNat` constraints are redundant.

### ISSUE THREAD

**Issue by YellPika** [expipiplus1/vector-sized#43]
Since `knownLength` exists, one can always obtain an instance of the `KnownNat` constraint from a `Vector`, so the constraint seems to be redundant for many functions that take `Vector`s as arguments (e.g. `Data.Vector.Sized.index`).

**Comment by mstksg** [comment]
Unfortunately, `knownLength` is not free; every invocation incurs a runtime cost.  So if we used it underneath every function taking `Vector`, it'd cause a lot of unnecessary runtime performance overhead :(

**Comment by mstksg** [comment]
Although it is true that we can remove `KnownNat` by shifting around some overhead to make things the same in the end, having the constraint does help give a more descriptive type signature, I think.  Performance might not be the main issue for many of the situations, but maybe just a descriptive type signature?

**Comment by YellPika** [comment]
The constraint isn't actually used in many functions. For example, `index` is implemented as:

```
index :: forall v n a. (KnownNat n, VG.Vector v a)
      => Vector v n a -> Finite n -> a
index (Vector v) i = v `VG.unsafeIndex` fromIntegral i
```

The constraint can simply be removed without adding any call to `knownLength`, so there will be no overhead.

How does the constraint make the signature more descriptive? If it's just an indicator that `n` is a type level natural, then I think it would be more appropriate to annotate the variable with a type (`n :: Nat`) instead of making the API unnecessarily restrictive.

**Comment by mstksg** [comment]
I do suppose this does not seem like an unreasonable change.  @expipiplus1 , any opinions?

**Comment by expipiplus1** [comment]
Closed with #44, Released as 1.0.3.0
